### PR TITLE
add config.SetMkldnnCacheCapacity api for mkldnn cache clear strategy

### PR DESCRIPTION
**Pull request by luotao1**
1. add `config.SetMkldnnCacheCapacity(xxx)` api for mkldnn cache clear strategy, xxx means the `mkldnn_cache_capacity` for difference input shape.
2. simplify the `TEST(Analyzer_MM_DNN, mkldnn_cache_clear)` with the new api, and add output compare between no cache strategy and using cache strategy.
3. add the `TEST(Analyzer_MM_DNN, mkldnn_shape_blob_size)` using `AnalysisPredictor::ZeroCopyRun`, since the new api will reset the `cur_mkldnn_session_id` to default.

related #18549

### CODE REVIEW

**Comment by luotao1** [comment]
@LeoZhao-Intel Please take a review!

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 1]
```diff
@@ -185,10 +185,49 @@ bool AnalysisPredictor::PrepareExecutor() {
   return true;
 }
 
+void AnalysisPredictor::MkldnnPreSet(const std::vector<PaddleTensor> &inputs) {
+#ifdef PADDLE_WITH_MKLDNN
+  VLOG(2) << "AnalysisPredictor::Run get_cur_mkldnn_session_id="
+          << platform::get_cur_mkldnn_session_id();
+  // In cache clearing mode.
+  if (config_.mkldnn_cache_capacity_ > 0) {
```
Comment (LeoZhao-BJ) [inline paddle/fluid/inference/api/analysis_predictor.cc:193]: There is confusion which one decides cache clearing mode, session_id or cache_capacity?
I think session_id is better, cache_capacity is just one property of session_id.

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 2]
```diff
@@ -185,10 +185,49 @@ bool AnalysisPredictor::PrepareExecutor() {
...
+    VLOG(2) << "Set input shape=" << ss.str();
+    platform::set_cur_input_shape_str(ss.str());
+  }
+#endif
+}
+
+void AnalysisPredictor::MkldnnPostReset() {
+#ifdef PADDLE_WITH_MKLDNN
+  // In cache clearing mode.
+  if (config_.mkldnn_cache_capacity_ > 0) {
```
Comment (LeoZhao-BJ) [inline paddle/fluid/inference/api/analysis_predictor.cc:215]: same comment with last one, which one decides cache clearing mode?

**Review by LeoZhao-BJ — COMMENTED** [review]
luotao, can you think if we can define mkldnn_config_ like mkldnn_quantizer_config_ does?

now we have 2 settings, session_id and capacity, can we put them into a config to reduce api change times?

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 1 above)
Comment (luotao1) [reply paddle/fluid/inference/api/analysis_predictor.cc:193]: `config_.mkldnn_cache_capacity_` is specified by the users, but `session_id` is used internally.
> which one decides cache clearing mode

Thus, for users, `config_.mkldnn_cache_capacity_` decides the cache clearing mode. 

In next PRs, I will move `paddle::platform::set_cur_mkldnn_session_id()`, `platform::set_cur_input_shape_cache_capacity(0)`,`platform::set_cur_input_shape_str("")` into `MKLDNNDeviceContext` class function. Or create a `MKLDNNDeviceContext::Reset()` function. i.e
```
void MKLDNNDeviceContext::Reset() {
  MKLDNNDeviceContext::set_cur_mkldnn_session_id(
        MKLDNNDeviceContext::kMKLDNNSessionID_Default);
    MKLDNNDeviceContext::set_cur_input_shape_cache_capacity(0);
    MKLDNNDeviceContext::set_cur_input_shape_str("");
}
```
And in `analysis_predictor.cc`, we will call
```
if (config_.mkldnn_cache_capacity_ > 0) {
  auto &pool = platform::DeviceContextPool::Instance();
  auto *dev_ctx = dynamic_cast<platform::MKLDNNDeviceContext *>(
      pool.Get(platform::CPUPlace()));
  dev_ctx->Reset();
}
```

**Comment by luotao1** [comment]
> luotao, can you think if we can define mkldnn_config_ like mkldnn_quantizer_config_ does?

It's maybe a little complex for users and business lines, since most of them can use only one line `config.EnableMKLDNN()` to trigger mkldnn ON or OFF. 

> now we have 2 settings, session_id and capacity, can we put them into a config to reduce api change times?

For inference api, we only have `capacity` one api. 
- I will move these setting into `MKLDNNDeviceContext` class in next PRs. Such as, `MKLDNNDeviceContext::SetCacheCapacity(int cache_capacity, std::string shape_str)`.
- In `analysis_predictor.cc`, we will not see `session_id` any more.

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 1 above)
Comment (LeoZhao-BJ) [reply paddle/fluid/inference/api/analysis_predictor.cc:193]: so you mean from user perspective, there are only 2 modes, one is normal mode with default cache_capacity setting, the other is cache clearing mode with explicit cache_capacity setting, session_id, shape_str are just internal variables used for implementation. right?

Comment (luotao1) [reply paddle/fluid/inference/api/analysis_predictor.cc:193]: Yes, you are right.

Comment (LeoZhao-BJ) [reply paddle/fluid/inference/api/analysis_predictor.cc:193]: LGTM

**Comment by luotao1** [comment]
@Superjomn Could you review the new `config.SetMkldnnCacheCapacity` api?

File: paddle/fluid/inference/analysis/argument.h [hunk 3]
```diff
@@ -148,6 +148,7 @@ struct Argument {
   // Pass a set of op types to enable its mkldnn kernel
   DECL_ARGUMENT_FIELD(mkldnn_enabled_op_types, MKLDNNEnabledOpTypes,
                       std::unordered_set<std::string>);
+  DECL_ARGUMENT_FIELD(mkldnn_cache_capacity, MkldnnCacheCapacity, int);
```
Comment (Superjomn) [inline paddle/fluid/inference/analysis/argument.h:152]: Add comment to describe this field.

Comment (luotao1) [reply paddle/fluid/inference/analysis/argument.h:152]: Done

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**